src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -926,13 +944,23 @@ case object SELFDESTRUCT extends OpCode(0xff, 1, 0, _.G_selfdestruct) {
   }
 
   protected def varGas[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): BigInt = {
+    val isValueTransfer = state.ownBalance > 0
+
     val (refundAddr, _) = state.stack.pop
     val refundAddress = Address(refundAddr)
 
-    if (state.config.chargeSelfDestructForNewAccount && !state.world.accountExists(refundAddress))
-      state.config.feeSchedule.G_newaccount
-    else
-      0
+    def postEip161CostCondition: Boolean =
+        state.config.chargeSelfDestructForNewAccount &&
+        isValueTransfer &&
+        state.world.isDead(refundAddress)
 
+    def preEip161CostCondition: Boolean =
+      state.config.chargeSelfDestructForNewAccount && !state.world.accountExists(refundAddress)
+
+
+    if (state.config.noEmptyAccounts)
+      if (postEip161CostCondition) state.config.feeSchedule.G_newaccount else 0
+    else
+      if (preEip161CostCondition) state.config.feeSchedule.G_newaccount else 0
```

Comment:
So... `if (noEmptyAccounts && postEIP161 || preEIP161) G_newaccount else 0` ?

Also has this logic has gotten quite more complex it might be desirable to share it between `varGas` and `gasExtra`

Reply:
I have tried the condition you propose but it doesn't work (blockchain test fails). it is because these conditions are separate i.e we don't want to check `preEIP161` if we are `postEIP161`. There are some cases where `noEmptyAccounts && postEIP161 ` will be false when we post EIP161, but 
`preEIP161` will be true and we would charge an incorrect fee.

Reply:
Pardon, that should be `if (noEmptyAccounts && postEIP161 || !noEmptyAccount && preEIP161) G_newaccount else 0` - which is a bit less concise but at least we don't repeat the branches.

Also my argument about sharing (not repeating) in `varGas` and `gasExtra` still stands.

Reply:
This condition is little less readable for me but we can go with it.

This condition (not really the same but similar) is used in `gasExtra` of `CallOp` and in `varGas` of `SelfDestruct`. You propose to share it between opCodes ?

Reply:
Sorry, I misread. I thought it was `CallOp#varGas`. Nevermind